// mocoolka-fp/lib/object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StrMap_1 = require("./StrMap");
var Array_1 = require("./Array");
var lodash_1 = require("lodash");
exports.get = lodash_1.get;
exports.has = lodash_1.has;
exports.mapKeys = lodash_1.mapKeys;
exports.mapValues = lodash_1.mapValues;
exports.pick = lodash_1.pick;
exports.pickBy = lodash_1.pickBy;
exports.merge = lodash_1.merge;
exports.moveAndFillRightObject = function (a) {
    var keys = Object.keys(a);
    var values = Array_1.moveAndFillRight(Object.values(a));
    return StrMap_1.fromFoldable(Array_1.array)(Array_1.zip(keys, values), function (existing, b) { return existing; }).value;
};
exports.withDefaults = function (defaults) { return function (props) {
    if (props === void 0) { props = {}; }
    return Object.assign({}, defaults, props);
}; };
exports.withDefaultsF = function (defaults) {
    return function (C) { return function (props) {
        return C(exports.withDefaults(defaults)(props));
    }; };
};
exports._get = function (a) { return function (b) { return lodash_1.get(b, a); }; };
/* export const _get = <T>(propName: TManyPropertyName) =>
(a:  object):  T => get(a, propName); */
