// mocoolka-fp/lib/ReaderEither.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var readerT = require("./ReaderT");
var Either_1 = require("./Either");
var readerTEither = readerT.getReaderT(Either_1.either);
exports.URI = 'ReaderEither';
var ReaderEither = /** @class */ (function () {
    function ReaderEither(run) {
        this.run = run;
    }
    ReaderEither.prototype.map = function (f) {
        return new ReaderEither(readerTEither.map(f, this.run));
    };
    ReaderEither.prototype.of = function (b) {
        return of(b);
    };
    ReaderEither.prototype.ap = function (fab) {
        return new ReaderEither(readerTEither.ap(fab.run, this.run));
    };
    ReaderEither.prototype.ap_ = function (fb) {
        return fb.ap(this);
    };
    ReaderEither.prototype.chain = function (f) {
        return new ReaderEither(readerTEither.chain(function (a) { return f(a).run; }, this.run));
    };
    return ReaderEither;
}());
exports.ReaderEither = ReaderEither;
var map = function (fa, f) {
    return fa.map(f);
};
var of = function (a) {
    return new ReaderEither(readerTEither.of(a));
};
var ap = function (fab, fa) {
    return fa.ap(fab);
};
var chain = function (fa, f) {
    return fa.chain(f);
};
var readerTask = readerT.ask(Either_1.either);
exports.ask = function () {
    return new ReaderEither(readerTask());
};
var readerTasks = readerT.asks(Either_1.either);
exports.asks = function (f) {
    return new ReaderEither(readerTasks(f));
};
exports.local = function (f) { return function (fa) {
    return new ReaderEither(function (e) { return fa.run(f(e)); });
}; };
exports.fromEither = function (fa) {
    return new ReaderEither(function () { return fa; });
};
exports.alt = function (fx, fy) {
    return new ReaderEither(function (e) { return fx.run(e).alt(fy.run(e)); });
};
exports.readerEither = {
    URI: exports.URI,
    map: map,
    of: of,
    ap: ap,
    chain: chain,
};
