// mocoolka-fp/lib/TypeCompare.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var t = require("./Type");
var c = require("./Compare");
var compareBooleanParams = {
    eq: 'boolean',
    not: 'boolean',
};
var compareArrayParamsT = function (type) { return ({
    contains: type,
    contains_every: [type],
    contains_some: [type],
}); };
var compareStringParams = {
    eq: 'string',
    not: 'string',
    lt: 'string',
    gt: 'string',
    lte: 'string',
    gte: 'string',
    in: ['string'],
    contains: 'string',
    notContains: 'string',
    notIn: ['string'],
    startsWith: 'string',
    endsWith: 'string',
    notStartsWith: 'string',
    notEndsWith: 'string',
};
var compareNumberParams = function (type) { return ({
    eq: type,
    not: type,
    lt: type,
    gt: type,
    lte: type,
    gte: type,
}); };
exports.compareParams = {
    string: compareStringParams,
    boolean: compareBooleanParams,
    number: compareNumberParams('number'),
    dateFromISOString: compareNumberParams('dateFromISOString'),
    dateFromNumber: compareNumberParams('dateFromNumber'),
    integer: compareNumberParams('integer'),
};
exports.compareArrayParams = {
    string: compareArrayParamsT('string'),
    boolean: compareArrayParamsT('boolean'),
    number: compareArrayParamsT('number'),
    dateFromISOString: compareArrayParamsT('dateFromISOString'),
    dateFromNumber: compareArrayParamsT('dateFromNumber'),
    integer: compareArrayParamsT('integer'),
};
exports.compare = {
    string: c.stringCompare,
    boolean: c.booleanCompare,
    number: c.numberCompare,
    dateFromISOString: c.numberCompare,
    dateFromNumber: c.numberCompare,
    integer: c.numberCompare,
};
exports.getCompareFunction = function (type, name) {
    if (t.RBasicType.is(type)) {
        var a = exports.compare[type];
        var b = a[name];
        if (!b) {
            throw new Error("Name '" + name + "' not be found in compare type '" + type + "'.");
        }
        return b;
    }
    else {
        throw new Error("type '" + type + "' not be found.");
    }
};
