// mocoolka-fp/lib/Setoid.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("fp-ts/lib/Setoid"));
var Setoid_1 = require("fp-ts/lib/Setoid");
var predicate_1 = require("./predicate");
exports.basicTypeEqual = function (a, b) {
    if (predicate_1.isNil(a) && predicate_1.isNil(b)) {
        return true;
    }
    else {
        return Setoid_1.strictEqual(a, b);
    }
};
exports.basicTypeSetoid = {
    string: Setoid_1.setoidString,
};
exports.setoidBasicType = { equals: exports.basicTypeEqual };
/** @function */
exports.getArrayASetoid = function (S) {
    return {
        equals: function (a, b) {
            if (predicate_1.isArray(a) && predicate_1.isArray(b)) {
                return Setoid_1.getArraySetoid(S).equals(a, b);
            }
            else if (predicate_1.isArray(a) || predicate_1.isArray(b)) {
                return false;
            }
            else {
                return S.equals(a, b);
            }
        },
    };
};
