// mocoolka-fp/lib/string/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var function_1 = require("../function");
__export(require("./camelCase"));
__export(require("./concat"));
__export(require("./constants"));
__export(require("./crop"));
__export(require("./empty"));
__export(require("./escapeHtml"));
__export(require("./escapeRegExp"));
__export(require("./escapeUnicode"));
__export(require("./hyphenate"));
__export(require("./lowerCase"));
__export(require("./lpad"));
__export(require("./ltrim"));
__export(require("./quoteString"));
__export(require("./removeNonASCII"));
__export(require("./removeNonWord"));
__export(require("./repeat"));
__export(require("./replace"));
__export(require("./replaceAccents"));
__export(require("./rtrim"));
__export(require("./slugify"));
__export(require("./symbols"));
__export(require("./toString"));
__export(require("./trim"));
__export(require("./truncate"));
__export(require("./unCamelCase"));
__export(require("./unescapeHtml"));
__export(require("./unhyphenate"));
__export(require("./upperCase"));
var repeat_1 = require("./repeat");
var hyphenate_1 = require("./hyphenate");
exports.repeatSpace = function (a) { return repeat_1.repeat('', a); };
exports.camelCaseToArray = function (str) {
    return hyphenate_1.hyphenate(str).split('-');
};
/**
 * Checks if string ends with specified suffix.
 */
exports.endsWith = function (suffix) { return function (str) {
    return str.indexOf(suffix, str.length - suffix.length) !== -1;
}; };
exports.notEndsWith = function (substring) { return function_1.not(exports.endsWith(substring)); };
/**
 * Searches for a given substring
 * @since v0.1.0
 * @category String
 * @ts String -> String -> Boolean
 * @param {string} substring
 * @param {string} str
 * @return {boolean}
 */
exports.includes = function (str, substring, fromIndex) {
    if (fromIndex === void 0) { fromIndex = 0; }
    return str.indexOf(substring, fromIndex) !== -1;
};
var lodash_1 = require("lodash");
exports.upperFirst = lodash_1.upperFirst;
exports.lowerFirst = lodash_1.lowerFirst;
exports.contains = function (substring) { return function (str) {
    return str.indexOf(substring, 0) !== -1;
}; };
exports.ins = function (v) { return function (str) {
    return v.includes(str);
}; };
exports.notIn = function (a) { return function (str) {
    return !exports.ins(a)(str);
}; };
exports.notContains = function (substring) { return function_1.not(exports.contains(substring)); };
/**
 * Checks if string ends with specified suffix.
 */
exports.startsWith = function (suffix) { return function (str) {
    return str.indexOf(suffix, 0) !== -1;
}; };
exports.notStartsWith = function (substring) { return function_1.not(exports.startsWith(substring)); };
